### PR TITLE
Topic/ref on jacobian

### PR DESCRIPTION
**Pull request by aescande**
This is a technical PR preparing changes in topic/subvariables.

The main goal of the PR is to switch the return of `jacobian` and `JDot` from matrix with properties to Ref with properties.
Along the way I'm also introducing utility function to better check where memory allocations arise with Eigen, and use them a bit to improve the code in some places.

For the second part, the problem is that the boolean value used to check if a memory allocation is allowed appears as a static variable in a header only function. Consequently the TVM library is getting a copy of it for its compiled functions, but Eigen's functions or any TVM header only function used from an other library or executable will use an instance of the boolean value compiled within the other library/executable. The main idea is then to have functions in TVM that are half inline/half compiled and ensure that both instances of the boolean are synchronized.

### CODE REVIEW

File: src/utils/memoryChecks.cpp [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+#endif
+}
+
+void restore_is_malloc_allowed_()
+{
+#ifdef EIGEN_RUNTIME_NO_MALLOC
+  assert(malloc_is_allowed_override_.size() > 0
+         && "restore_is_malloc_allowed called too many times compared to override_is_malloc_allowed.");
+  if(Eigen::internal::is_malloc_allowed() != malloc_is_allowed_override_.top())
+    Eigen::internal::set_is_malloc_allowed(malloc_is_allowed_override_.top());
```
Comment (gergondet) [inline src/utils/memoryChecks.cpp:45]: Why do you check the current `is_malloc_allowed` state before?

**Review by gergondet — APPROVED** [review]
Just a few questions, but otherwise looks good, thanks for the short PR :+1:

File: include/tvm/scheme/internal/CompiledAssignment.h [hunk 2]
```diff
@@ -183,20 +186,25 @@ template<typename MatrixType>
...
-      Eigen::internal::set_is_malloc_allowed(true);
-      cache_ = M;
-      return cache_;
-      Eigen::internal::set_is_malloc_allowed(false);
+      // For the case where M = matrix * Constant, Eigen create a temporary for constant before evaluating the product
+      // We avoid that here, by doing the product by hand.
+      // This could be further optimized by taking into acount the WeightMult at once, and possibly without using a
+      // cache.
+      cache_ = M.lhs().rowwise().sum(); // TODO compare to a handmade loop suming the columns
+      cache_ *= M.rhs().functor().m_other;
```
[the remark this answers is not in the document]
Comment (aescande) [reply include/tvm/scheme/internal/CompiledAssignment.h:202]: > 
> 
> I'm a little confused here (and the similar code below), you mean that `cache_.noalias() = somematrix * constant;` will generate allocations?
>

Yes: deep within the product evaluation, `Eigen::VectorXd::Constant` is transformed into a temporary `Eigen::VectorXd`

> But I'm not sure I understand the operations you are doing in this case

We want to perform `M v`, where v is a vector with the form `v = c 1` with `c` a constant and `1` a vector of ones.
The product can be rewritten `M c 1 = c M 1`, and `M 1` corresponds to the sum of the columns of `M`.

I'll add some comments.

File: src/utils/memoryChecks.cpp (hunk 1 above)
Comment (aescande) [reply src/utils/memoryChecks.cpp:45]: It avoids calling the function, but I guess I just mimicked the way we did similar things before.

Comment (gergondet) [reply src/utils/memoryChecks.cpp:45]: Internally (at least on 3.3.4 but I guess this hasn't changed a lot between Eigen versions) `is_malloc_allowed` and `set_is_malloc_allowed` call the same function so you end up with two calls (sometimes) when one call is always enough

File: include/tvm/utils/memoryChecks.h [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+#ifdef NDEBUG
+inline bool is_malloc_allowed() { return true; }
+inline bool set_is_malloc_allowed(bool b) { return b; }
+inline void override_is_malloc_allowed(bool) {}
+inline void restore_is_malloc_allowed() {}
+#else
+TVM_DLLAPI bool is_malloc_allowed_();
+TVM_DLLAPI bool set_is_malloc_allowed_(bool allow);
+TVM_DLLAPI void override_is_malloc_allowed_(bool v);
+TVM_DLLAPI void restore_is_malloc_allowed_();
```
[the remark this answers is not in the document]
Comment (aescande) [reply include/tvm/utils/memoryChecks.h:23]: Done in 9c4703c3957f9eef036519a0c57e644a9ed5f622

File: include/tvm/scheme/internal/CompiledAssignment.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (aescande) [reply include/tvm/scheme/internal/CompiledAssignment.h:203]: Comments added in 9c4703c3957f9eef036519a0c57e644a9ed5f622

File: src/utils/memoryChecks.cpp (hunk 1 above)
Comment (aescande) [reply src/utils/memoryChecks.cpp]: I removed the `if` line in 9c4703c3957f9eef036519a0c57e644a9ed5f622

**Pull request merged**